Show fallback message in extensions marketplace when the store can't connect to the WCCOM API

andfinally opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

- When a store is unable to connect to the WooCommerce.com API to show extensions in the in-app marketplace, we'd like to show the user a message with a link to WCCOM. 
- This branch modifies `WC_Admin_Addons::render_featured` to check the response from the API's "featured" endpoint more carefully, and call a new `output_empty` method if it doesn't seem correct.
- Likewise, it changes `WC_Admin_Addons::get_extension_data` to return a `WP_Error` if the API's search endpoint doesn't return a good response. If the addons page template finds this error, it also calls `output_empty`.
- `output_empty` renders a simple error message.
- When a connection error happens, it does a `woocommerce_page_wc-addons_connection_error` action. This records a `wcadmin_extensions_view_connection_error` Tracks event.
- When a connection error happens on the featured page, the error message contains an extra line giving more detail about the error.

Closes https://github.com/Automattic/woocommerce.com/issues/11757.

### How to test the changes in this Pull Request:

1. Check out this branch on your local environment.
2. Go to to Extensions page `/wp-admin/admin.php?page=wc-addons`.
3. First, test the good case.
    1. Check that you see a list of extensions on this "featured" page.
    2. Search for a term like "shipping" and see the results.
    3. Search for a term that won't get any results.
    4. Browse to one of the categories and see there are extensions listed.
4. Now break the API request by downloading this [patch](https://github.com/woocommerce/woocommerce/files/7514631/breaking-marketplace-api.txt) and appying it with `git apply ~/Downloads/breaking-marketplace-api.txt`.
5. Repeat the steps from 3. You should see the message in each case. 
    1. On the featured page, the error message will have an extra line with the status code 404.
6. If you have access to the Tracks Live Viewer, note that after a few minutes, a [wcadmin_extensions_view_connection_error](https://mc.a8c.com/tracks/events/event.php?eventname=wcadmin_extensions_view_connection_error) event has been recorded.
7. Revert the changes from the patch.
8. Add this on line 163 of `wp-includes/Requests/Transport/cURL.php` on your test site. This forces the site to connect using an out of date TLS version.

```php
curl_setopt( $this->handle, CURLOPT_SSLVERSION, CURL_SSLVERSION_MAX_TLSv1_1 );
```

9. View the "featured" page `/wp-admin/admin.php?page=wc-addons`.  On this page only, the error message will have an extra line about the SSL error. 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

Update - Show fallback message when extensions marketplace cannot connect to WCCOM API.

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### Screenshots

#### Featured page - 404 error message

<img width="500" src="https://user-images.githubusercontent.com/1647564/141810780-b1cc5f6d-ca17-4554-924a-8f6dbacf128c.png">

#### Featured page – SSL error message

<img width="500" src="https://user-images.githubusercontent.com/1647564/141810753-1d6785a2-b635-48ad-83a4-8b8a51ef6d7a.png">

#### Category page

<img width="500" src="https://user-images.githubusercontent.com/1647564/141160799-d788bc36-46f5-43ec-a746-faba98c2874a.png">

andfinally committed:
Showing fallback message when the store can't connect to the WCCOM API.

andfinally committed:
Slightly tidier handling of bad responses in `render_featured`.

andfinally committed:
Added `woocommerce_page_wc-addons_connection_error` action, which records a Track event when the in-app marketplace fails to connect properly to the WCCOM API.

andfinally committed:
Added more informative error messages when API request fails on the in-app marketplace featured page.

andfinally commented:
Please note, we currently have a [request](https://woomarketingstudio.wordpress.com/2021/11/15/product-copy-for-marketplace-error/) open for the wording of the error message, so the text of that may change in the next day or two.

andfinally committed:
Updated error text.

andfinally commented:
Wording now updated in https://github.com/woocommerce/woocommerce/pull/31126/commits/299c73a50e56c0b38b6cd9a76c810a81617fd6ff.

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-addons.php:
```diff
@@ -87,17 +87,58 @@ public static function render_featured() {
+				return;
+			}
+
+			$response_code = intval( wp_remote_retrieve_response_code( $raw_featured ) );
```
Minor, but WP encourages type casting now: https://core.trac.wordpress.org/ticket/42918

andfinally replied:
Thanks, TIL! Changed in https://github.com/woocommerce/woocommerce/pull/31126/commits/1dfd1c14e8f7bdf9df6cc45d745d7c0f17ea4770.

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-addons.php:
```diff
@@ -144,9 +185,24 @@ public static function get_extension_data( $category, $term, $country ) {
+			return $raw_extensions;
+		}
+
+		$response_code = intval( wp_remote_retrieve_response_code( $raw_extensions ) );
```
Same here.

andfinally replied:
Changed in https://github.com/woocommerce/woocommerce/pull/31126/commits/1dfd1c14e8f7bdf9df6cc45d745d7c0f17ea4770.

peterfabian commented on plugins/woocommerce/includes/admin/views/html-admin-page-addons.php:
```diff
@@ -81,42 +81,46 @@ class="<?php echo $current_section === $section->slug ? 'current' : ''; ?>"
-						WC_Admin_Addons::output_search_promotion_block( $promotion );
+			<?php if ( '_featured' !== $current_section ) : ?>
+				<?php if ( is_wp_error( $addons ) ) : ?>
+					<?php WC_Admin_Addons::output_empty(); ?>
```
For sake of consistency and being more useful for support, I wonder if you folks didn't want to include some message also here (from `get_extension_data`).

andfinally replied:
Good point, thanks!

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-addons.php:
```diff
@@ -946,8 +1027,12 @@ public static function output() {
+				$addons = $extension_data;
+			} else {
+				$addons = $extension_data->products;
+			}
```
Stylistic nitpick: in the list of function calls and tertiary operators, this one looks a bit out of place.

andfinally replied:
Nice! I've changed this to a ternary in https://github.com/woocommerce/woocommerce/pull/31126/commits/1dfd1c14e8f7bdf9df6cc45d745d7c0f17ea4770.

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-addons.php:
```diff
@@ -87,17 +87,58 @@ public static function render_featured() {
+	}
+
+	public static function is_ssl_error( $error_message ) {
+		return ( false !== stripos( $error_message, 'cURL error 35' ) );
```
Is it possible that there are other cURL codes for SSL errors that we want to capture? Or are we interested in only this specific one?

andfinally replied:
We're mostly interested in catching this case, but I'm now outputting the cURL error in other cases too.

peterfabian reviewed (dismissed):
Tested, all worked fine, except I wasn't able to trigger the SSL error, but maybe that's because I'm using the Local environment that has different TLS settings.

Only minor nitpicks as comments, so feel free to merge if you don't agree with them.

Note to future self: Next time, disable uBlock when testing tracks ;) :D

peterfabian commented:
Ah, one more thing. I think I encountered this when the internet was cut off:
![Screenshot 2021-11-18 at 14 46 01](https://user-images.githubusercontent.com/2207451/142470665-f82a04eb-2525-42fa-910b-f7e500bb6704.png)

andfinally committed:
Addressing feedback.

- Replaced `intval` calls with faster type casting.
- Showing more informative error messages.
- Using ternary for better legibility in `WC_Admin_Addons::output`.
- Added `is_wp_error` check to avoid warning from `count` call in `html-admin-page-addons.php`.
- Always showing error message as detail in error output.
- Showing error detail message above "To start growing your business, head over to WooCommerce.com" text, instead of below.

andfinally commented:
Thanks for all the great improvements @peterfabian 🙌 I've added them in in https://github.com/woocommerce/woocommerce/pull/31126/commits/1dfd1c14e8f7bdf9df6cc45d745d7c0f17ea4770 and rebased.

peterfabian reviewed (approved):
Looks good, thanks for the updates.

peterfabian commented:
@andfinally Do you want to wait to potentially refine the wording or is it good to be merged?

andfinally commented:
The wording's good to go now, thanks @peterfabian!
